extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,315 @@
+import com.google.common.collect.Maps;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import io.druid.collections.LoadBalancingPool;
+import io.druid.collections.ResourceHolder;
+import io.druid.collections.StupidResourceHolder;
+import net.jpountz.lz4.LZ4Exception;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.ShardedJedis;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+import redis.clients.jedis.exceptions.JedisException;
+import redis.clients.util.Hashing;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class RedisCache implements Cache
+{
+  private static final Logger log = new Logger(RedisCache.class);
+
+  private final Supplier<ResourceHolder<ShardedJedis>> supplier;
+  private final String prefix;
+  private final long expiration;
+  private final RedisTranscoder transcoder;
+  private final AtomicLong hitCount = new AtomicLong(0);
+  private final AtomicLong missCount = new AtomicLong(0);
+  private final AtomicLong timeoutCount = new AtomicLong(0);
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    final Supplier<ResourceHolder<ShardedJedis>> clientSupplier;
+
+    if (config.getPoolSize() > 1) {
+      clientSupplier = new LoadBalancingPool<>(
+          config.getPoolSize(),
+          new Supplier<ShardedJedis>()
+          {
+            @Override
+            public ShardedJedis get()
+            {
+              return new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH);
+            }
+          }
+      );
+    } else {
+      clientSupplier = Suppliers.<ResourceHolder<ShardedJedis>>ofInstance(
+          StupidResourceHolder.create(new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH))
+      );
+    }
+
+    return new RedisCache(clientSupplier, config);
+  }
+
+  public RedisCache(Supplier<ResourceHolder<ShardedJedis>> supplier, RedisCacheConfig config)
+  {
+    this.supplier = supplier;
+    prefix = config.getPrefix();
+    expiration = config.getExpiration();
+    transcoder = new RedisTranscoder();
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    byte[] result = null;
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      final byte[] serialized = jedis.get(itemKey);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (IOException ex) {
+      throw Throwables.propagate(ex);
```

Comment:
we should increase errorCount, log a warning and return null instead of propagating the error so the query can still complete even if the cache is having problems.